### PR TITLE
Fix race condition

### PR DESCRIPTION
**Pull request by methane**
like #21, but `PostRawData()` doesn't block while reconnecting.

### CODE REVIEW

**Comment by repeatedly** [comment]
What is the blocker for merging this?

**Comment by repeatedly** [comment]
Having race conditin bug is not good for users so we should fix the problem.
I don't know the status of #19 and #21.
If merge this patch, can we close #19 and #21 ?

File: fluent/fluent.go [hunk 1]
```diff
@@ -248,44 +265,45 @@ func e(x, y float64) int {
...
-}
+func (f *Fluent) send() error {
+	f.muconn.Lock()
+	defer f.muconn.Unlock()
 
-func (f *Fluent) send() (err error) {
 	if f.conn == nil {
 		if f.reconnecting == false {
-			f.mu.Lock()
 			f.reconnecting = true
```
Comment (mattn) [inline fluent/fluent.go]: Is this atomic?

Comment (methane) [inline fluent/fluent.go]: Fluent.reconnecting is guarded by f.muconn.

Comment (mattn) [inline fluent/fluent.go]: Ah, sorry. I miss it.

**Comment by tagomoris** [comment]
Now I have a chance to try this patch in an environment which seems to be facing with the race condition problem. Wait for the confirmation in that environment.

File: fluent/fluent.go [hunk 2]
```diff
@@ -156,26 +163,23 @@ func (f *Fluent) PostWithTime(tag string, tm time.Time, message interface{}) err
...
+	f.PostRawData(data)
+	return nil
 }
 
 func (f *Fluent) PostRawData(data []byte) {
-	f.mu.Lock()
-	f.pending = append(f.pending, data...)
-	f.mu.Unlock()
+	if err := f.appendBuffer(data); err != nil {
+		// TODO: return err. Since it may change public API, doesn't fix for now.
```
Comment (tagomoris) [inline fluent/fluent.go]: How about adding another method which does same thing with `PostRawData`, but returns error?
In that way, we don't break public API, but get this library enable to handle errors for appending data into buffer.

Comment (tagomoris) [inline fluent/fluent.go]: I believe that almost users don't use `PostRawData` directly.

Comment (methane) [inline fluent/fluent.go]: OK, I've marked it as deprecated.

**Comment by tagomoris** [comment]
I confirmed that fluent-logger-golang with this change doesn't panic and returns errors with the dummy logger client and Fluentd plugin/configuration set of #29.

Returned errors are:

1. "write tcp 127.0.0.1:36746->127.0.0.1:24224: write: connection reset by peer"
1. "fluent#send: can't send logs, client is reconnecting" for a while
1. "Buffer full" at last

**Comment by tagomoris** [comment]
LGTM. Time to merge... I'll update error message about buffer full after merge.

**Pull request merged**